mlflow/models/evaluation/default_evaluator.py

```diff
@@ -522,6 +522,18 @@ def _get_aggregate_metrics_values(metrics):
 }
 
 
+def _extract_chat_message_response_or_prediction(prediction: list) -> str:
+    # Handle case when ChatMessage sequence is returned by the predict func
+    if (
+        prediction is not None
```

Comment:
The first check isn't needed. 

Can this logic be enhanced a bit to explicitly only fetch for dictionaries that contain the "content" key so that unexpected parsing issues don't occur? 

If the prediction function returns multiple ChatMessage constructs, is it guaranteed that the last message is the only one of worthwhile context?

Reply:
* Done
* Done
* No, but it seemed better than the below options

Options
1. Expose a `pull_last_message` bool parameter in mlflow.evaluate. This is bad because it exposes a niche parameter in a very public API
2. Expose a `chat_message_parser` for each chat the ChatMessage sequence. This is overkill
3. ~6 other bad ideas
4. Something else I'm missing

The current logic is a bit brittle but resolves the bug. Am I overindexing on the fact that these APIs will be redesigned and we just want a quick fix?

Reply:
I'm not sure if solving this in the general evaluator is the right thing to do. Imo the fact that `mlflow.evaluate` does not handle list of chat messages is not a bug, but rather one of many cases where model output does not comply with the format that the particular evaluator expects. For example, `AgentExecutor` outputs a dictionary of `{"input": xxx, "outputs": yyy}` which is also not a chat message. It is not realistic to handle all different possible output from LlamaIndex workflow, DSPy, etc.

We are actually going thorough large re-design of this API and potentially split the model prediction part into different API. This should give a way for customers to adjust the prediction format into what evaluator expects.

For the time being, you can use [Evaluating with a Custom Function](https://mlflow.org/docs/latest/llms/llm-evaluate/index.html#evaluating-with-a-custom-function) to parse the agent output. While it adds one more step, I believe it's clearer than baking hidden data transformation inside this giant monolith API🤔.

Reply:
Yep agreed. @B-Step62 let's use this as an evidence-based example for the CUJ justification for splitting the API :) I think it's a compelling demonstration of the interface struggles that exist where even the workaround's functionality is somewhat cumbersome and frustrating.